### PR TITLE
[recipe] fix: add support for Jenkins job to update recipes

### PR DESCRIPTION
**Pull request by Wayonb**
problem: currently recipes are updated manually
solution: create Jenkins to update the recipes

Fix build issue with RocksDB and mongo-c-driver.
Will need to build the packages and tests with the catapult client.

### CODE REVIEW

File: scripts/CatapultRecipeUpdater.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+RECIPES = RECIPES_REPOSITORIES.keys()
+
+
+def dispatch_subprocess(command_line, cwd=None, handle_error=True):
+	print(' '.join(command_line))
+	result = subprocess.run(command_line, check=False, cwd=cwd, capture_output=True)
+	if handle_error and 0 != result.returncode:
+		raise subprocess.SubprocessError(f'command failed with exit code {result.returncode}\n{result}')
+
+	output = result.stdout.decode('utf-8') if 0 == result.returncode else result.stderr.decode('utf-8')
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: could refactor decode outside of if/else

File: scripts/CatapultRecipeUpdater.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+	return output, result.returncode
+
+
+class CatapultRecipesUpdater:
+	def __init__(self, source_path):
+		self.source_path = Path(source_path).absolute()
+
+	@staticmethod
+	async def _initialize_conan():
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: would move this out as free function and call directly in main

File: scripts/CatapultRecipeUpdater.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+		async with ClientSession(raise_for_status=True) as session:
+			async with session.get(url) as response:
+				response_json = await response.json()
+				if 200 != response.status:
+					raise Exception(f'failed to get latest release for {owner}/{repo} {response_json}')
+				return re.sub(r'[^\d\.]', '', response_json['tag_name'])
+
+	@staticmethod
+	async def _read_yaml_file(filepath):
+		with open(filepath, 'r') as file:
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: rt

File: scripts/CatapultRecipeUpdater.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+		print(f'checking recipe {recipe} {current_version} -> {latest_version}')
+		if semver.compare(latest_version, current_version) > 0:
+			print(f'{recipe} has new version: {latest_version}')
+			recipe_update[recipe] = (current_version, latest_version)
+
+		return recipe_update
+
+	async def get_available_update(self, recipes):
+		results = await asyncio.gather(*[self._get_update_if_available(recipe) for recipe in recipes])
+		recipe_with_updates = dict(ChainMap(*results))
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: why are you using ChainMap?
might be cleaner to return tuple of (recipe, current_version, latest_version) or None, and then build here 

vs returning bunch of dicts with one key and merging

File: scripts/CatapultRecipeUpdater.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+		return recipe_with_updates
+
+	async def _update_conandata_sha(self, recipe, filepath, version):
+		config = await self._read_yaml_file(filepath)
+		url = config['sources'][version]['url']
+		with tempfile.TemporaryDirectory() as tmpdir:
+			download_file = f'{tmpdir}/{recipe}.{version}.tar.gz'
+			dispatch_subprocess(['curl', '-LJ', url, '--output', f'{download_file}'])
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: just `download_file` ?

File: scripts/CatapultRecipeUpdater.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+	async def upload_conan_package(self, recipes_versions):
+		for recipe, versions in recipes_versions:
+			_, new_version = versions
+			recipe_path = self.source_path / f'{recipe}/all'
+			dispatch_subprocess(
+				['conan', 'upload', f'{recipe}/{new_version}@nemtech/stable', '--remote=nemtech'],
+				cwd=recipe_path,
+				handle_error=True
+			)
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: consider refactoring as this is same as previous with only difference being conan command

File: scripts/CatapultRecipeUpdater.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+async def main():
+	parser = argparse.ArgumentParser(description='Recipes updater for catapult')
+	parser.add_argument('--commit-title', help='commit title')
+	parser.add_argument('--recipes', choices=RECIPES, help='recipes to update', default=RECIPES)
+	parser.add_argument('--recipes-path', help='path to the recipes', required=True)
+	parser.add_argument('--upload', help='upload recipes to conan', action='store_true')
+	args = parser.parse_args()
+
+	recipes_updater = CatapultRecipesUpdater(args.recipes_path)
+	recipes_to_update = await recipes_updater.get_available_update(args.recipes)
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: get_available_update*s ?

File: scripts/CatapultRecipeUpdater.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+				yaml.dump(config, output_file, sort_keys=False)
+
+	async def _update_recipe_files(self, recipe, versions, recipes_versions):
+		update_recipe_files_descriptor = [
+			r'{recipe}/config.yml',
+			r'{recipe}/all/conandata.yml',
+			r'{recipe}/all/test_package/CMakeLists.txt',
+			r'{recipe}/all/conanfile.py',
+		]
+		dependency_map = {'mongo-cxx-driver': 'mongo-c-driver', 'cppzmq': 'zeromq'}
```
Comment (Jaguar0625) [inline scripts/CatapultRecipeUpdater.py]: suggestion - might be cleaner to move out any specific package knowledge outside of CatapultRecipeHlper into helper function/classes [also applies to 'zeromq' checks elsewhere].

File: .github/jenkinsfile/recipeUpdater.groovy [hunk 9]
```diff
@@ -0,0 +1,7 @@
+gitPullRequestPipeline {
+	dockerImageName = 'symbolplatform/build-ci:cpp-ubuntu-lts'
+	branchName = 'fix/update_catapult_recipe'
```
Comment (Jaguar0625) [inline .github/jenkinsfile/recipeUpdater.groovy]: i guess should be changed to `main` before merge?

**Review by Jaguar0625 — APPROVED** [review]

File: .github/jenkinsfile/recipeUpdater.groovy (hunk 9 above)
Comment (Wayonb) [reply .github/jenkinsfile/recipeUpdater.groovy]: This is the pull request branch name. I will rename it for clarity.
The default base branch, ``main`` for this repo, is configured in the Jenkins job and can be overridden when the job is run manually.

File: scripts/CatapultRecipeUpdater.py (hunk 4 above)
Comment (Wayonb) [reply scripts/CatapultRecipeUpdater.py]: ChainMap holds a list of dicts but will simply and refactor it 👍🏾

File: scripts/CatapultRecipeUpdater.py (hunk 5 above)
Comment (Jaguar0625) [reply scripts/CatapultRecipeUpdater.py]: do you need `f'{tmp_source_filepath}` or would just `tmp_source_filepath` work below (2x)?

**Review by Jaguar0625 — APPROVED** [review]
👍

**Pull request merged**